### ISSUE TITLE
Context Object for initialization

### ISSUE THREAD

**Issue by milleruntime** [apache/accumulo#1177]
I am not sure if this should be done or not so I am opening an issue for discussion/feedback/design ideas.  

InitializeContext would be the object to represent the special internal state of Accumulo, before/during its initialization.  This became troublesome when I started touching the VolumeManager. 

<pre>
server/base/src/main/java/org/apache/accumulo/server/InitializeContext.java
public class InitializeContext extends ServerContext {
  private VolumeManager fs;

  public InitializeContext(SiteConfiguration siteConf, Configuration hadoopConf) throws IOException {
    super(siteConf, hadoopConf);
    fs = VolumeManagerImpl.get(siteConf, hadoopConf);
}
</pre>

Most of the changes would be with server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java.  I was thinking maybe this context object could have a simple boolean state to check whether Accumulo has been initialized or not.  Thoughts?

**Comment by keith-turner** [comment]
What do you see as the benefit of this change?

**Comment by milleruntime** [comment]
> What do you see as the benefit of this change?

The same benefits that came with the creation of the other context objects.  Less static code and better testing.

**Comment by keith-turner** [comment]
>  Less static code and better testing.

I don't know enough about the specifics to give any useful feedback, but the goal seems great.

**Comment by ctubbsii** [comment]
I remember chatting with somebody about creating something like a UtilityContext for the various server-side utilities, for the same reason. Do you think the Initialize command is sufficiently distinct that it needs its own, or do you think it could use a more general UtilityContext?

**Comment by milleruntime** [comment]
> I remember chatting with somebody about creating something like a UtilityContext for the various server-side utilities, for the same reason. Do you think the Initialize command is sufficiently distinct that it needs its own, or do you think it could use a more general UtilityContext?

I am not sure.  I would hate to have too many Context objects, that would be confusing.  The nice thing about having an IntializedContext is that it would capture the state only during the initialize process.  Having a UtilityContext seems a little too broad.  Perhaps a "ServerUtilContext" would be better in at least you know where the context is coming from.  But that just seems like what ServerContext would do.  What else would a UtilityContext object handle differently than ServerContext?

**Comment by ctubbsii** [comment]
I was definitely thinking just a "ServerUtilContext", not a more general utility context. The main thing is that, unlike the ClientContext, it would assume access to the server-side configuration, but unlike the ServerContext, it wouldn't do things that are needed for server's only (crypto service, unique name allocator, etc.).

It's probably best to start with just an InitializeContext, and not worry about a general purpose one unless we need it.

**Comment by milleruntime** [comment]
OK I will give this a stab.

**Comment by milleruntime** [comment]
I think a good first step for this would be to refactor `Initialize` into multiple classes. This may give us a type of context object that would be easier to maintain but I don't think we want to extend `ServerContext`. The things we can do during initialization are very limited and if anything `ClientContext` may be able to extend an "IntializeContext", if there is enough shared logic to make it a part of the class hierarchy.

**Comment by ctubbsii** [comment]
I think the bulk of the motivation for having it share class hierarchy with ServerContext is that it uses a lot of the utility code we wrote elsewhere that accepts a ServerContext, particularly for working with ZooKeeper stuff. But, it doesn't have to do that if it doesn't make sense to do so.

Another hiccup that makes this frustrating is that ServerInfo uses ZooKeeper and HDFS objects in some of its constructors to do instanceId and instanceName lookups. This creates a chicken-egg bootstrap problem a bit. Ideally, ServerInfo should be an object representation of the configuration properties, and that is used to construct a ServerContext. It really shouldn't have stateful objects that do work like ZooKeeper connections. Those kinds of objects should exist in ServerContext. So, it should go something like: `SiteConfiguration` :arrow_right: `ServerInfo` :arrow_right: `ServerContext` (a composition something like: `new ServerContext(new ServerInfo(new SiteConfiguration()))`). However, because ServerInfo provides the instance ID, and because it may need to look up the instance ID from HDFS, or from an instance name to look it up in ZooKeeper, it has to create worker objects. This would be a lot simpler if the instance name was unique already and served as the instance ID as well. This would prevent being able to rename an instance, but if the instance name was included in the server config file (and the volumes had a sanity check to ensure it matched), we wouldn't need to do any lookups for the instance ID in `ServerInfo`. Currently, though, we have both instanceName and instanceId for Initialize, and neither for regular server bootups. Clients are expected to do a lookup of the instanceId from an instanceName, and servers are expected to look up from HDFS using `instance.volumes`. It's all very needlessly complex, but I think we're stuck with some of it for awhile.

**Comment by milleruntime** [comment]
Yeah the chicken-egg bootstrap problem was my main problem with including it in the class hierarchy. My first step was to break up `Initialize` into multiple class objects, organizing all the static code so we can better see what can be reused/shared.
